### PR TITLE
Locale support, db connection closing.

### PR DESCRIPTION
**Pull request by andy1ne0**
- Initial commit - locale support. See LocaleExample for a demonstration of usage. 
- Properly terminate database connection - to make sure any pending queries are executed, and to also avoid any sort of errors. 

Notes: The strings directory structure is temporary, until an official plugin structure is determined. Put it there for now to keep it all in one place. 

Relates to: 
- CraftTogether/Discussion-Issues#12.

### CODE REVIEW

File: src/main/java/org/crafttogether/strings/LocaleManager.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.ArrayList;
+import java.util.Optional;
+
+/**
+ * Locale manager, used to get the translated version of a string or register a new translation object.
+ * @author Andrew Petersen
+ * @version 0.1
+ */
+
+public class LocaleManager {
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleManager.java]: Codacy Issue found: [Utility classes should not have a public or default constructor.](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875248/issues/source?bid=3277702&fileBranchId=3283068#l12)

File: src/main/java/org/crafttogether/strings/LocaleLanguage.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * @version 0.1
+ */
+
+public enum LocaleLanguage {
+
+    ENGLISH(0), GERMAN(1), FRENCH(2), PIRATE(3); // TODO Register more languages as needed.
+
+    private final int id;
+
+    LocaleLanguage(int i){
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleLanguage.java]: Codacy Issue found: [Missing a Javadoc comment.](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875247/issues/source?bid=3277702&fileBranchId=3283068#l15)

File: src/main/java/org/crafttogether/strings/LocaleLanguage.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        this.id = i;
+    }
+
+    public int getId(){
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleLanguage.java]: Codacy Issue found: [Missing a Javadoc comment.](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875247/issues/source?bid=3277702&fileBranchId=3283068#l19)

File: src/main/java/org/crafttogether/strings/LocaleExample.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+*/
+
+public class LocaleExample extends LocaleString {
+
+    /**
+     * The English translation should be a constant. Otherwise, there is the risk that translations won't be returned.
+     */
+    public static final String ENGLISH_TRANSLATION = "Hello, and welcome to the CraftTogether Server!";
+
+    public LocaleExample() {
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleExample.java]: Codacy Issue found: [Missing a Javadoc comment.](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875246/issues/source?bid=3277702&fileBranchId=3283068#l16)

File: src/main/java/org/crafttogether/strings/LocaleString.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+public abstract class LocaleString {
+
+    private HashMap<LocaleLanguage, String> translations = new HashMap<>();
+
+    /**
+     * Get the English translation of the string.
+     * @return The English translation.
+     */
+    public String getEnglishString() {
+        return translations.get(LocaleLanguage.ENGLISH);
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleString.java]: Codacy Issue found: [Reference to instance variable 'translations' needs "this.".](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875249/issues/source?bid=3277702&fileBranchId=3283068#l20)

File: src/main/java/org/crafttogether/strings/LocaleString.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * The English string is required in the constructor as the default, should no other options be available.
+     * This should be a constant.
+     * @param englishString The English translation of the string - the default if a translation is not provided.
+     */
+    public LocaleString(String englishString){
+        translations.put(LocaleLanguage.ENGLISH, englishString);
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleString.java]: Codacy Issue found: [Reference to instance variable 'translations' needs "this.".](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875249/issues/source?bid=3277702&fileBranchId=3283068#l29)

File: src/main/java/org/crafttogether/strings/LocaleString.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * @param lang The language of the new translation.
+     * @param translated The string to assign to the said language.
+     * @return Whether the translation was successfully registered or not.
+     */
+    public boolean registerTranslation(LocaleLanguage lang, String translated){
+        if(translations.keySet().contains(lang)){
+            return false;
+        } else {
+            translations.put(lang, translated);
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleString.java]: Codacy Issue found: [Reference to instance variable 'translations' needs "this.".](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875249/issues/source?bid=3277702&fileBranchId=3283068#l47)

File: src/main/java/org/crafttogether/strings/LocaleString.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * Get the translation of this string.
+     * @param lang The language of the translation to return.
+     * @return The translated string. Defaults to English if the specified language is not registered.
+     */
+    public String getTranslation(LocaleLanguage lang){
+        if(translations.containsKey(lang)) {
+            return translations.get(lang);
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleString.java]: Codacy Issue found: [Reference to instance variable 'translations' needs "this.".](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875249/issues/source?bid=3277702&fileBranchId=3283068#l59)

File: src/main/java/org/crafttogether/strings/LocaleString.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    }
+
+    /**
+     * Get the translation of this string.
+     * @param lang The language of the translation to return.
+     * @return The translated string. Defaults to English if the specified language is not registered.
+     */
+    public String getTranslation(LocaleLanguage lang){
+        if(translations.containsKey(lang)) {
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleString.java]: Codacy Issue found: [Reference to instance variable 'translations' needs "this.".](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875249/issues/source?bid=3277702&fileBranchId=3283068#l58)

File: src/main/java/org/crafttogether/strings/LocaleString.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    public abstract void registerAllTranslations();
+
+    /**
+     * @param lang The language of the new translation.
+     * @param translated The string to assign to the said language.
+     * @return Whether the translation was successfully registered or not.
+     */
+    public boolean registerTranslation(LocaleLanguage lang, String translated){
+        if(translations.keySet().contains(lang)){
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleString.java]: Codacy Issue found: [Reference to instance variable 'translations' needs "this.".](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529875249/issues/source?bid=3277702&fileBranchId=3283068#l44)

File: src/main/java/org/crafttogether/strings/LocaleManager.java (hunk 1 above)
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleManager.java]: Codacy Issue found: [Utility classes should not have a public or default constructor.](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529877643/issues/source?bid=3277702&fileBranchId=3283068#l12)

File: src/main/java/org/crafttogether/strings/LocaleLanguage.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * @version 0.1
+ */
+
+public enum LocaleLanguage {
+
+    ENGLISH(0), GERMAN(1), FRENCH(2), PIRATE(3); // TODO Register more languages as needed.
+
+    private final int id;
+
+    LocaleLanguage(int i){
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleLanguage.java]: Codacy Issue found: [Missing a Javadoc comment.](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529877642/issues/source?bid=3277702&fileBranchId=3283068#l15)

File: src/main/java/org/crafttogether/strings/LocaleManager.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
```
Comment (theminecoder) [inline src/main/java/org/crafttogether/strings/LocaleManager.java]: Codacy Issue found: [Class LocaleManager should be declared as final.](https://www.codacy.com/app/CraftTogether/CraftCore/file/2529887045/issues/source?bid=3277702&fileBranchId=3283068#l12)

**Comment by GingerGeek** [comment]
Please only one change per pull request. :smile:

**Comment by theminecoder** [comment]
Also can we not have example classes with `main(String[])` methods? If you are going to have an example please have it in the code where it would be used properly.

**Comment by andy1ne0** [comment]
Aight, I'll get rid of that/delete the extra change this afternoon and reopen

**Comment by andy1ne0** [comment]
Created a separate PR with the appropriate changes: #15. Locking conversation on this PR.

**Pull request closed without merging**